blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/Containers.java

```diff
@@ -15,7 +15,133 @@
+                final int end = limit < 0 ? -1 : (start + limit);
+                Iterator<? extends T> iterator;
+                private T next;
+                int index = 0;
+                int iterableIndex = -1;
+                {
+                    // primer
+                    nextIterator();
+                    moveNext();
+                }
+                private void nextIterator() {
+                    iterableIndex++;
+                    if (iterableIndex < iterables.length) {
+                        iterator = iterables[iterableIndex].iterator();
+                    } else {
+                        iterator = null;
+                    }
+                }
+                private void moveNext() {
+                    // Skip until we hit our starting index
+                    while (index < start) {
+                        if (iterator == null) {
+                            return;
+                        }
+                        if (iterator.hasNext()) {
+                            iterator.next();
+                            index++;
+                        } else {
+                            nextIterator();
+                        }
+                    }
+
+                    // keep adding until we hit the limit or run out of items
+                    if (index < end || end < 0) {
+                        while (iterator != null && !iterator.hasNext()) {
+                            nextIterator();
+                            if (iterator == null) {
+                                return;
+                            }
+                        }
+
+                        next = iterator.next();
+                        index++;
+                    } else {
+                        iterableIndex = iterables.length;
+                        iterator = null;
+                    }
+                }
+
+                @Override
+                public boolean hasNext() {
+                    return iterator != null;
+                }
+
+                @Override
+                public T next() {
+                    try {
+                        return next;
+                    } finally {
+                        moveNext();
+                    }
+                }
+
+                @Override
+                public void remove() {
+                    throw new RuntimeException("remove() not implemented.");
+                }
+            };
+        }
+    }
+
+    /**
+     * Simple method to provide a combination of {@link Iterable} with the ability to implement a specific
+     * {@link #get(String)} method in a subclass. This will also iterate as efficiently as possible
+     * through the provided iterators to provide paged data - e.g. requesting page 0-9 will only
+     * iterate through the first 10 items, in the order of the provided {@link Iterable}
+     */
+    public abstract static class IterableContainer<T extends Resource> extends Container<T> {
```

Comment:
@vivek there are two scenarios that come to mind:
1. often a user will receive a deep-link to the run details screen for a build that is queued. The URL in this case contains the "expectedBuildNumber" value. With the current setup the UI would need to fetch against /runs/:id and receive a 404. It could then try to fetch the item from /queue but it's very bad to have 404's if we can avoid them.
2. in other scenarios the UI will receive SSE for a job when it becomes queued. We may want to fetch data about the queue item via /queue/:id but frequently the build will transition to a "running" state even in the short amount of time between when the SSE is received and when the fetch is made. In this case the UI would receive a 404 and need to fetch from /runs/:id.

IMO from a cleanliness standpoint and for ease of debugging it's desirable not to have 404's.